src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::credentials::errors::CredentialsError;
+use base64::Engine;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+#[allow(dead_code)]
+type Result<T> = std::result::Result<T, CredentialsError>;
+
+/// Token Exchange grant type for a sts exchange.
+#[allow(dead_code)]
+pub const TOKEN_EXCHANGE_GRANT_TYPE: &str = "urn:ietf:params:oauth:grant-type:token-exchange";
+/// Refresh Token Exchange grant type for a sts exchange.
+#[allow(dead_code)]
+pub const REFRESH_TOKEN_GRANT_TYPE: &str = "refresh_token";
+/// TokenType for a sts exchange.
+#[allow(dead_code)]
+pub const ACCESS_TOKEN_TYPE: &str = "urn:ietf:params:oauth:token-type:access_token";
+/// JWT TokenType for a sts exchange.
+#[allow(dead_code)]
+pub const JWT_TOKEN_TYPE: &str = "urn:ietf:params:oauth:token-type:jwt";
+
+/// Handles OAuth2 Secure Token Service (STS) exchange.
+/// Reference: https://datatracker.ietf.org/doc/html/rfc8693
+pub struct STSHandler {
+    client: reqwest::Client,
+}
+
+#[allow(dead_code)]
+impl STSHandler {
+    pub fn new() -> Self {
+        let client = reqwest::Client::new();
+        Self { client }
+    }
+
+    /// performs the token exchange using a refresh token flow with
+    /// the provided [RefreshAccessTokenRequest] information.
+    pub async fn refresh_access_token(
+        &self,
+        req: RefreshAccessTokenRequest,
+    ) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+        params.insert("grant_type", REFRESH_TOKEN_GRANT_TYPE.to_string());
+        params.insert("refresh_token", req.refresh_token);
+
+        self.execute(req.url, req.authentication, req.headers, params)
+            .await
+    }
+
+    /// performs an oauth2 token exchange with the provided [ExchangeTokenRequest] information.
+    pub async fn exchange_token(&self, req: ExchangeTokenRequest) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+
+        params.insert("grant_type", TOKEN_EXCHANGE_GRANT_TYPE.to_string());
+        params.insert("requested_token_type", ACCESS_TOKEN_TYPE.to_string());
+
+        params.insert("subject_token", req.subject_token);
+        params.insert("subject_token_type", req.subject_token_type);
+
+        if !req.scope.is_empty() {
+            params.insert("scope", req.scope.join(" "));
+        }
+
+        if let Some(audience) = req.audience {
+            params.insert("audience", audience);
+        }
+        if let Some(resource) = req.resource {
+            params.insert("resource", resource);
+        }
+        if let Some(actor_token) = req.actor_token {
+            params.insert("actor_token", actor_token);
+        }
+        if let Some(actor_token_type) = req.actor_token_type {
+            params.insert("actor_token_type", actor_token_type);
+        }
```

Comment:
Not sure, but maybe:

```suggestion
       [
           ("audience", req.audience),
           ("resource", req.resource),
           ("actor_token", req.actor_token),
           ("actor_token_type", req.actor_token_type),
        ]
        .into_iter()
        .filter_map(|(name, v)| v.map(|s| (name, s)))
        .for_each(|(name, v)| params.insert(name, v);
```